gson/src/main/java/com/google/gson/JsonPrimitive.java

```diff
@@ -287,11 +287,16 @@ public boolean equals(Object obj) {
           : this.getAsNumber().longValue() == other.getAsNumber().longValue();
     }
     if (value instanceof Number && other.value instanceof Number) {
-      double a = getAsNumber().doubleValue();
-      // Java standard types other than double return true for two NaN. So, need
-      // special handling for double.
-      double b = other.getAsNumber().doubleValue();
-      return a == b || (Double.isNaN(a) && Double.isNaN(b));
+      if (value instanceof BigDecimal && other.value instanceof BigDecimal) {
+        // Uses compareTo to ignore scale of values, e.g. `0` and `0.00` should be considered equal
+        return this.getAsBigDecimal().compareTo(other.getAsBigDecimal()) == 0;
```

Comment:
As side note / for future reference: I originally proposed using `compareTo` when this code was also comparing `BigDecimal` with non-`BigDecimal`, see https://github.com/google/gson/pull/2364#discussion_r1155158943.

However, using `compareTo` instead of `equals` here has another advantage: It is _transitive_ (as required by [`Object.equals`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#equals(java.lang.Object))).
> (`x equals y` and `y equals z`) implies `x equals z`

Now let
- `x = new JsonPrimitive(new BigDecimal("0"))`
- `y = new JsonPrimitive(0.0d)`
- `z = new JsonPrimitive(new BigDecimal("0.00"))`

Using `compareTo` makes sure that `equals` is transitive. Whereas using `BigDecimal.equals` here would cause: `x equals y` and `y equals z` (because they fall back to `double` comparison), but `x !equals z`, which would not be transitive.